servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ProxyTunnel.java

```diff
@@ -105,72 +108,71 @@ int connectCount() {
-                    final OutputStream out = socket.getOutputStream();
-                    out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
-                    out.flush();
-
-                    final InputStream cin = clientSocket.getInputStream();
-                    Future<Void> f = executor.submit(() -> {
-                        copyStream(out, cin);
-                        return null;
-                    });
-                    copyStream(clientSocket.getOutputStream(), in);
-                    f.get(); // wait for the copy of proxy client input to server output to finish copying.
+    private void handleRequest(final Socket socket, final String initialLine) throws IOException, ExecutionException,
+            InterruptedException {
+        if (initialLine.startsWith(CONNECT_PREFIX)) {
+            final int end = initialLine.indexOf(' ', CONNECT_PREFIX.length());
+            final String authority = initialLine.substring(CONNECT_PREFIX.length(), end);
+            final String protocol = initialLine.substring(end + 1);
+            final int colon = authority.indexOf(':');
+            final String host = authority.substring(0, colon);
+            final int port = Integer.parseInt(authority.substring(colon + 1));
+
+            try (Socket clientSocket = new Socket(host, port)) {
+                connectCount.incrementAndGet();
+                final OutputStream out = socket.getOutputStream();
+                out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
+                out.flush();
+
+                Future<Void> f = executor.submit(() -> {
+                    try {
+                        copyStream(out, clientSocket.getInputStream());
+                    } finally {
+                        socket.shutdownOutput();
+                        clientSocket.shutdownInput();
+                    }
+                    return null;
+                });
+                try {
+                    copyStream(clientSocket.getOutputStream(), socket.getInputStream());
+                } finally {
+                    clientSocket.shutdownOutput();
+                    socket.shutdownInput();
+                    f.get(); // Wait for copy of proxy client input to server output to finish.
                 }
-            } else {
-                throw new RuntimeException("Unrecognized initial line: " + initialLine);
             }
-        } finally {
-            in.close();
+        } else {
+            throw new IllegalArgumentException("Unrecognized initial line: " + initialLine);
         }
+        // The Socket is closed outside the scope of this method.
     }
 
     private static void copyStream(final OutputStream out, final InputStream cin) throws IOException {
-        try {
-            int b;
-            while ((b = cin.read()) >= 0) {
-                out.write(b);
-            }
+        int read;
+        final byte[] bytes = new byte[1024];
```

Comment:
If you prefer to buffer, consider using a smaller size, like 8 or 16 bytes. Requests in the test are tiny. The intend of the proxy is to fragment data as much as possible.